### PR TITLE
Use formatting string for custom formattable

### PR DESCRIPTION
**Pull request by SimonTouchtech**
**What issue does this PR address?**
#1396

**Does this PR introduce a breaking change?**
No.

**Please check if the PR fulfills these requirements**
- [x] The commit follows our [guidelines](https://github.com/serilog/serilog/blob/dev/CONTRIBUTING.md)
- [x] Unit Tests for the changes have been added (for bug fixes / features)

**Other information:**
This is my first PR to Serilog and I don't have much experience with the more complex parts of the library, so I hope this doesn't break anything.

It looked like the issue appeared in the last line of `PropertyValueConverter.CreatePropertyValue` where `new ScalarValue(value.ToString())` was used as a fallback. By calling `ToString()` and thus converting to a string the custom formatting capabilities are lost. Therefore I added a special case for `IFormattable` just before the `ToString()` call. The reason for putting it last was to not interfere with any destructuring.

It is arguable whether a custom implementation of `IFormattable` could be considered "scalar" as the type can potentially be very complex. The current behavior of `ScalarValue` worked as requested though, so I decided to use it. An alternative approach could be to do something like `FormattableValue` to make the naming a bit more fitting for complex custom `IFormattable` implementations.

**Changes:**
- The `PropertyValueConverter` now returns a `ScalarValue` without modification for `IFormattable` objects

### CODE REVIEW

**Comment by nblumhardt** [comment]
Thanks! Unfortunately, there's a bit of a subtlety around `ScalarValue`. Whatever it wraps needs to be thread-safe, at least for formatting, and this is hard to guess/guarantee for arbitrary `IFormattable` types, hence the current `ToString()` capturing.

**Comment by SimonTouchtech** [comment]
Thanks for the feedback @nblumhardt! That is indeed unfortunate, as that makes this problem significantly harder than I initially thought.

I might look more into this later in the week, but for now I don't see a clear path to solve it. As you mentioned, thread safety can't be guaranteed for an arbitrary `IFormattable` type, I also don't see a way to create another implementation of `LogEventPropertyValue` (or an extension of `ScalarValue`) that would enforce thread-safety. Therefore I don't see a way to "keep" the formatting capability after the `LogEventPropertyValue` has been created.

My current best idea would be to somehow use the formatting capability before it's "removed" by the `ToString()` capturing. One fairly straight forward way of doing that would be something like

```csharp
if (value is IFormattable formattable)
    return new ScalarValue(formattable.ToString(format));

return new ScalarValue(value.ToString());
```

I see two rather big issues with such a solution though:

1. Formatting/rendering is done in a place where it doesn't belong. It usually happens in the `MessageTemplateRenderer` but now it also happens in the creation of the property value. However, it could be argued that even the existing `ToString()` capturing is a form of rendering, so the only difference would be that the captured value will be more "correct".
2. The `format` is not available in `CreatePropertyValue` so the method signature would have to be changed. As the method is public that would not only be a breaking change, but it would also arguably be a breaking change that doesn't make much sense (see 1).

I guess it would be possible to not change the method signature of `CreatePropertyValue` by abusing the fact that it takes an `object` and we could do something like the following for `IFormattable` implementations.

```csharp
_valueConverter.CreatePropertyValue(new Foo(value, propertyToken.Format), propertyToken.Destructuring));
```

And then have some sort of special case for `Foo`. But IMO that'd be even worse than changing the method signature...

In summary, I don't see any solution that I think would be good. I'll try to have another look later this week, but unless I come up with something better than what I wrote above I'll likely close this PR.
